src/com/perl5/lang/perl/idea/refactoring/PerlRefactor.java

```diff
@@ -0,0 +1,420 @@
+        first = false;
+      }
+      else {
+        builder.append(", ");
+      }
+      builder.append(vd.name.replaceAll("[%@]", "\\$"));
+    }
+    builder.append(") = @_;");
+    return builder.toString();
+  }
+
+  /**
+   * @param newParameters
+   * @return built string of method parameters
+   */
+  public String buildMethodParameters(@Nullable ArrayList<String> newParameters) {
+    StringBuilder builder = new StringBuilder();
+    if (!this.className.isEmpty()) {
+      builder.append("    my ").append(this.className).append(" = shift;\n");
+    }
+    ArrayList<String> parameters;
+
+    if (newParameters != null) {
+      parameters = newParameters;
+    }
+    else {
+      parameters = this.parameters;
+    }
+    builder.append("    my (");
+    boolean first = true;
+    for (String s : parameters) {
+      if (first) {
+        first = false;
+      }
+      else {
+        builder.append(", ");
+      }
+      builder.append(s.replaceAll("[%@]", "\\$"));
+    }
+    builder.append(") = @_;");
+
+    this.methodParameters = builder.toString();
+    return this.methodParameters;
+  }
+
+  private void parseLocalVariables() {
+
+    for (String var : this.scalarVariables) {
+      String reg1 = String.format("\\s*my\\s*\\%s\\s*[=;(]", var);
+      String reg2 = String.format("\\s*my\\s*\\(.*?\\%s.*?\\)", var);
+      String reg3 = String.format("(?:for|foreach)\\s+my\\s*\\%s\\s*\\(", var);
+
+      Pattern pattern1 = Pattern.compile(reg1);
+      Pattern pattern2 = Pattern.compile(reg2);
+      Pattern pattern3 = Pattern.compile(reg3);
+
+      if (var.matches("(?:\\$\\d+$|\\$[ab]$)")) {
+        this.localScalarVariables.add(var);
+      }
+      else if (pattern1.matcher(codeSelection).find() || pattern2.matcher(codeSelection).find()) {
+        this.localScalarVariables.add(var);
+
+        // Skip loop variables
+        if (pattern3.matcher(codeSelection).find()) {
+          this.loopScalarVariables.add(var);
+        }
+      }
+    }
+
+    for (String var : this.arrayVariables) {
+      var = var.replace("$", "@");
+
+      String reg1 = String.format("\\s*my\\s*\\%s\\s*[=;(]", var);
+      String reg2 = String.format("\\s*my\\s*\\(.*?\\%s.*?\\)", var);
+      Pattern pattern1 = Pattern.compile(reg1);
+      Pattern pattern2 = Pattern.compile(reg2);
+
+      if (pattern1.matcher(codeSelection).find() || pattern2.matcher(codeSelection).find()) {
+        localArrayVariables.add(var);
+      }
+    }
+
+    for (String var : this.hashVariables) {
+      var = var.replace("$", "%");
+      String reg1 = String.format("\\s*my\\s*\\%s\\s*[=;(]", var);
+      String reg2 = String.format("\\s*my\\s*\\(.*?\\%s.*?\\)", var);
+      Pattern pattern1 = Pattern.compile(reg1);
+      Pattern pattern2 = Pattern.compile(reg2);
+
+      if (pattern1.matcher(codeSelection).find() || pattern2.matcher(codeSelection).find()) {
+        this.localHashVariables.add(var);
+      }
+    }
+  }
+
+  private void parseVariables() {
```

Comment:
This is wrong approach. Unlike some others editors or ide this should not be done with regular expressions.
There is a psi tree and you should work with it. Traverse tree, collect elements, resolve them. Understand which are outer and which are not. 

I'd suggest to look for examples in other plugins or idea itself.

Reply:
This will take some time for me to figure out.

Reply:
true.